### PR TITLE
Stream to distribution example

### LINKED ISSUE #874: [Examples] - Polish and needed checks
# Apps - Small Review

[protocol-monorepo/examples/employment-based-loan/contracts at dev · superfluid-finance/protocol-monorepo · GitHub](https://github.com/superfluid-finance/protocol-monorepo/tree/dev/examples/employment-based-loan/contracts)

- No checking if caller is host or with the right agreement.
- If not needed don’t save CFA and host if using CFALib (already point this in older PR: https://github.com/superfluid-finance/protocol-monorepo/pull/772)

[protocol-monorepo/BudgetNFT.sol at dev · superfluid-finance/protocol-monorepo · GitHub](https://github.com/superfluid-finance/protocol-monorepo/blob/dev/examples/budget-nft/contracts/BudgetNFT.sol)

- Not using CFALib
- Why Migrations.sol file?

[protocol-monorepo/examples/nftbillboard-userdata at dev · superfluid-finance/protocol-monorepo · GitHub](https://github.com/superfluid-finance/protocol-monorepo/tree/dev/examples/nftbillboard-userdata)
- Remove .github folder
- If not needed don’t save CFA and host if using CFALib (already talked about this in older PR: https://github.com/superfluid-finance/protocol-monorepo/pull/772)

[protocol-monorepo/DividendRightsToken.sol at dev · superfluid-finance/protocol-monorepo · GitHub](https://github.com/superfluid-finance/protocol-monorepo/blob/dev/examples/rewards-distribution-token/contracts/DividendRightsToken.sol)

- No checking if caller is host or with the right agreement.
- If not needed don’t save HOST if using IDALib

### PR DESCRIPTION
**Pull request by jtriley2p**

### CODE REVIEW

File: examples/stream-in-distribute-out/contracts/base/StreamInDistributeOut.sol
Comment (hellwolf) [inline examples/stream-in-distribute-out/contracts/base/StreamInDistributeOut.sol:77]:
```
require(_isCFAv1(agreementClass), "LotterySuperApp: only CFAv1 supported");
```

should always check like this, please check our existing example code base too

Comment (ngmachado) [reply examples/stream-in-distribute-out/contracts/base/StreamInDistributeOut.sol:77]: After checking the rest of examples, we are very inconsistent about this very important check. Opening a new issue to keep track.

Comment (jtriley2p) [reply examples/stream-in-distribute-out/contracts/base/StreamInDistributeOut.sol:77]: This can not revert, because `approveSubscription` triggers either the `afterAgreementCreated` or `afterAgreementUdpated` (depending on if the subscription exists).
https://github.com/superfluid-finance/protocol-monorepo/blob/089010c1403a930c392435ec173dc37c76689cfc/packages/ethereum-contracts/contracts/agreements/InstantDistributionAgreementV1.sol#L327

This prevents anyone from approving subscriptions if the callbacks revert when the agreement is the IDA.

---

We are checking at the start of every callback and only returning the `ctx` if it's not the CFA.

**Comment by ngmachado** [comment]
@JoshuaTrujillo15 sync this PR with dev branch. After that we can merge it!

**Review by ngmachado — APPROVED** [review]

**Pull request merged**
